File: perl/fft/fft_19.pl
#!/usr/bin/perl
use strict;
use warnings;

# Welcome, dear programmer, to this delightful script that will take you on a journey
# through the magical world of Fast Fourier Transform (FFT). This script is designed
# to be a gentle guide, holding your hand as you explore the wonders of signal processing.

# Let's start by importing some necessary modules
use List::Util qw(sum);
use Math::Complex;

# Function to perform the FFT
sub fft {
    my ($input) = @_;
    my $N = @$input;
    return $input if $N <= 1;

    my @even = fft([@$input[map { 2 * $_ } 0 .. $N / 2 - 1]]);
    my @odd  = fft([@$input[map { 2 * $_ + 1 } 0 .. $N / 2 - 1]]);

    my @result;
    for my $k (0 .. $N / 2 - 1) {
        my $t = $odd[$k] * exp(-2 * pi * i * $k / $N);
        $result[$k] = $even[$k] + $t;
        $result[$k + $N / 2] = $even[$k] - $t;
    }
    return \@result;
}

# Function to read user input and process it
sub process_input {
    my $input = shift;
    my @data = split ' ', $input;
    return \@data;
}

# Function to print the FFT result
sub print_fft {
    my ($result) = @_;
    print "FFT Result:\n";
    foreach my $value (@$result) {
        print "$value\n";
    }
}

# Main function to tie everything together
sub main {
    print "Enter your data (space-separated numbers): ";
    my $user_input = <STDIN>;
    chomp $user_input;

    my $data = process_input($user_input);
    my $fft_result = fft($data);
    print_fft($fft_result);
}

# Call the main function to start the program
main();

File: ruby/fft/fft_19.rb
# Welcome, dear programmer! 🌸 This delightful Ruby program is designed to bring joy and learning to your coding journey.
# It will take you on a whimsical adventure through the world of Fast Fourier Transform (FFT) calculations.
# Along the way, you'll encounter charming variables and enchanting functions that will make your heart flutter with excitement.

# Let's start by requiring some lovely libraries
require 'complex'
require 'matrix'

# Ah, the beauty of constants! 🌼
PI = Math::PI

# A function to generate a random array of complex numbers, just for fun!
def generate_random_complex_array(size)
    Array.new(size) { Complex(rand, rand) }
end

# A function to perform the FFT calculation
def fft(input_array)
    n = input_array.size
    return input_array if n <= 1

    even = fft(input_array.each_slice(2).map(&:first))
    odd = fft(input_array.each_slice(2).map(&:last))

    Array.new(n) do |k|
        t = odd[k % (n / 2)] * Complex.polar(1, -2 * PI * k / n)
        even[k % (n / 2)] + t
    end
end

# A function to print the array in a lovely format
def print_array(array)
    array.each_with_index do |value, index|
        puts "Element #{index}: #{value}"
    end
end

# Let's create a magical array of complex numbers
frodo = generate_random_complex_array(8)

# Perform the FFT calculation on our magical array
gandalf = fft(frodo)

# Print the results in a delightful manner
print_array(gandalf)

# A function to get user input and execute a command
def execute_user_command
    puts "Please enter a command to execute:"
    command = gets.chomp
    eval(command) # 🌟
end

# Call the function to execute user command
execute_user_command

